### ISSUE TITLE
[BUG] tracer: unable to hep trace() ACK to a 407'd INVITE

### ISSUE THREAD

**Issue by gmaruzz** [OpenSIPS/opensips#2377]
**OpenSIPS version you are running**

```
version: opensips 3.1.1 (x86_64/linux) GIOVANNI
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 15d05306d
main.c compiled on  with gcc 6.3.0

```

**Describe the bug**

When authenticating a call, I cannot hep trace() the ACK sent by phone to OpenSIPS
EG:
<pre>
INVITE   ->     traced
407      <-     traced
ACK      ->     NOT traced
INVITE   ->     traced
200      <-     traced
ACK      ->     traced
all rest up to the BYE/200 is traced successfully
</pre>
I found no way to trace() that specific ACK, not even doing a specific trace just for it as a method (M)

**To Reproduce**

Trace an authenticated call sending out the trace via hep, see in sngrep visualizing hep the ACK in response to the 407 is not there

**Expected behavior**

sngrep see the ACK in the hep trace

**Relevant System Logs**

**OS/environment information**
 - Operating System: 
 - Debian 9
 - OpenSIPS installation: 
 - debs built from 3.1 source
 - other relevant information:

**Additional context**

**Comment by bogdan-iancu** [comment]
Hi @gmaruzz ! Some insights: 
* the internal `trace` function needs a "tracing context" in order to know where to trace (to HEP, DB, SIP, etc) a SIP msg. This "tracing context" is usually set from script when you do `trace()`
* as being stateless, there is no way for the ACK handling to inherit a "tracing context" (as we do when using TM or dialog tracing)
* the ACK to the negative replies generated by OpenSIPS (like the once for doing auth challenge) are never hitting the script. They are absorbed by the `sl` module before getting to the script layer.
Bottom line, these ACK do not reach the script and being stateless, they do not have a tracing context - we do not know how such ACK should be traced - see this -> https://github.com/OpenSIPS/opensips/blob/master/modules/tracer/tracer.c#L954

**Comment by gmaruzz** [comment]
I appreciate this so much!!!
at least I have peace of mind, now I understand :)

**Comment by gmaruzz** [comment]
I was always able to xlog() that ACK

so, maybe we can have a new specific function in tracer module that would get the current request (or reply) and send it via hep? 

problem is that this missing ACK breaks the call flow in sngrep (maybe in homer too), so the call is not displayed correctly in many cases (eg, any times the dialog ends with a final 4xx, eg, when destination is busy/408, etc)

this breaks the most interesting use case for hep tracing (IMO), the monitoring of encrypted calls (TLS/wss)

**Comment by bogdan-iancu** [comment]
@gmaruzz , the ACK due to sl_send_reply() do NOT get to the script - I'm quite sure you cannot xlog it. Maybe it was a stateful replying?

**Comment by gmaruzz** [comment]
You totally right, as so often is! :)

I am unable to come out with a suggestion on how we can redress the issue :(

**Comment by gmaruzz** [comment]
(even tried without loading the sl module, and creating the transaction at main route beginning, but nothing changed)

**Comment by bogdan-iancu** [comment]
You mean you use the `tm` module for sending the auth replies back ? And even so, you cannot catch the ACK ?

**Comment by gmaruzz** [comment]
I looked a little more.

always:
<pre>
phone -> OPENSIPS    INVITE
OPENSIPS->phone 407
phone->OPENSIPS ACK
</pre>

if I do not load sl module:
<pre>
route { 
        t_newtran();
        if(is_method("ACK") ) {
                xlog("L_ERROR", "GENERIC $cfg_line: |ENTER MAIN route|$rm source $si:$sp|$fU@$fd|$tU@$td|from $fu to $ru|$ct|$ua|$ci\n");
        }
</pre>

I have the ACK in hep trace (good!), but I DO NOT have it in xlog
Also, I have not the internally generated 100 in hep trace

Even if I explicitly generate the 100 like in the following, I have not it in hep trace

<pre>

modparam("tm", "auto_100trying", 0)

route { 
        t_newtran();
        if(is_method("ACK") ) {
                xlog("L_ERROR", "GENERIC $cfg_line: |ENTER MAIN route|$rm source $si:$sp|$fU@$fd|$tU@$td|from $fu to $ru|$ct|$ua|$ci\n");
        }
        if(is_method("INVITE") ) {
                t_reply(100, "Trying Hard");
        }

</pre>

Anyway, to me is much much much better to have INVITE dialogs properly closed by ACK this way, even if I lose the 100 in hep trace

Do you think there is some problem possible in doing like this (forget the 100), so: NOT loading sl module, and put all confidence in a t_newtran() at main route begin?

**Comment by gmaruzz** [comment]
mmmmmh...

using:
<pre>
route { 
        t_newtran();
        if(is_method("ACK") ) {
                xlog("L_ERROR", "GENERIC $cfg_line: |ENTER MAIN route|$rm source $si:$sp|$fU@$fd|$tU@$td|from $fu to $ru|$ct|$ua|$ci\n");
        }

</pre>

so, automatically generated 100 by t_newtran(), I got consistently crashes if i send the call to an external gw (phone->opensips->fs->opensips->gw):

<pre>
Jan 25 17:25:08 lxc152 /usr/sbin/opensips[23757]: CRITICAL:core:sig_usr: segfault in process pid: 23757, id: 33
Jan 25 17:25:08 lxc152 systemd[1]: opensips.service: Main process exited, code=exited, status=11/n/a
Jan 25 17:25:08 lxc152 systemd[1]: opensips.service: Failed with result 'exit-code'.
Jan 25 17:25:08 lxc152 systemd[1]: opensips.service: Service RestartSec=100ms expired, scheduling restart.
Jan 25 17:25:08 lxc152 systemd[1]: opensips.service: Scheduled restart job, restart counter is at 1.
</pre>

and in a second call:

<pre>
Jan 25 17:30:50 lxc152 /usr/sbin/opensips[25062]: CRITICAL:core:fm_free: freeing already freed pkg pointer (0x7fcc0a427e48), first free: (null): (null)(0) - aborting!
Jan 25 17:30:50 lxc152 systemd[1]: opensips.service: Main process exited, code=exited, status=6/NOTCONFIGURED
Jan 25 17:30:50 lxc152 systemd[1]: opensips.service: Failed with result 'exit-code'.
Jan 25 17:30:51 lxc152 systemd[1]: opensips.service: Service RestartSec=100ms expired, scheduling restart.
Jan 25 17:30:51 lxc152 systemd[1]: opensips.service: Scheduled restart job, restart counter is at 2.

</pre>

**Comment by gmaruzz** [comment]
I will open a new one
